File: frontend/src/types/ticket.ts
import type { TicketSlaState } from './sla'

/** Catálogo único de 10 estados, compartido por Ticket y Tarea/Subtarea (spec 009) — una
 * Tarea puede transicionar a cualquiera de ellos sin restricción de secuencia (ver
 * ticketService.changeStatus). */
export type TicketStatus =
  | 'nuevo' | 'pre_analisis' | 'contacto' | 'en_analisis' | 'en_ejecucion'
  | 'en_pruebas' | 'pendiente_usuario' | 'resuelto' | 'cerrado' | 'cancelado'

export type TicketType = 'incident' | 'evolutive' | 'preventive'
export type Priority = 'critical' | 'high' | 'medium' | 'low'
export type Severity = 's1' | 's2' | 's3' | 's4'
export type EscalationLevel = 'n1' | 'n2' | 'n3' | 'n4'

export const STATUS_LABELS: Record<TicketStatus, string> = {
  nuevo: 'Nuevo', pre_analisis: 'Pre-Análisis', contacto: 'Contacto',
  en_analisis: 'En Análisis', en_ejecucion: 'En Ejecución', en_pruebas: 'En Pruebas',
  pendiente_usuario: 'Pendiente de Usuario', resuelto: 'Resuelto',
  cerrado: 'Cerrado', cancelado: 'Cancelado',
}

export const TICKET_TYPE_LABELS: Record<TicketType, string> = {
  incident: 'Incidente', evolutive: 'Evolutivo', preventive: 'Preventivo',
}

export const PRIORITY_LABELS: Record<Priority, string> = {
  critical: 'Crítica', high: 'Alta', medium: 'Media', low: 'Baja',
}

export const SEVERITY_LABELS: Record<Severity, string> = {
  s1: 'S1', s2: 'S2', s3: 'S3', s4: 'S4',
}

export interface EntityRef {
  id: string
  name: string
}

export interface ResourceRef {
  id: string
  full_name: string
}

export interface TicketListItem {
  id: string
  ticket_number: string
  record_type_id: string
  ticket_type: TicketType
  title: string
  status: TicketStatus
  status_label: string
  priority: Priority
  severity: Severity
  escalation_level: EscalationLevel
  client: EntityRef | null
  project: EntityRef | null
  assignee: ResourceRef | null
  estimated_resolution_minutes: number | null
  /** Nombre de la Lista resuelto (spec 009 — reemplaza el texto libre de la spec 008). */
  list_name: string | null
  list_id: string | null
  /** Nombre resuelto del catálogo tipo de registro ("Ticket" | "Tarea") — usado para el tag
   * del Kanban y de "Mis Tareas" sin round-trip adicional (spec 009). */
  record_type: 'Ticket' | 'Tarea'
  /** Si no es null, este registro es una Subtarea (Nivel 5) de la Tarea indicada. */
  parent_task_id: string | null
  created_at: string
  /** Resumen de SLA (Fase 4, spec 014) — solo `phase`/`status`, sin el detalle completo. */
  sla: Pick<TicketSlaState, 'phase' | 'status'>
}

/** "125" min → "2h 05m"; null → '—'. Usado en el tablero Kanban. */
export function formatMinutes(minutes: number | null): string {
  if (minutes == null) return '—'
  const h = Math.floor(minutes / 60)
  const m = minutes % 60
  return h > 0 ? `${h}h ${String(m).padStart(2, '0')}m` : `${m}m`
}

export type CommentType =
  | 'asignado' | 'pre_analisis' | 'confirmacion_atencion' | 'solicitud_informacion'
  | 'termina_analisis' | 'solicitud_cierre' | 'respuesta_usuario'
  | 'descripcion_solucion' | 'comentario_interno' | 'cancelacion'

export interface TicketAttachment {
  id: string
  filename: string
  content_type: string
  size_bytes: number
}

export interface TicketComment {
  id: string
  comment_type: CommentType
  comment_type_label: string
  visibility: 'internal' | 'external'
  body: string
  author_id: string
  is_automatic: boolean
  attachments: TicketAttachment[]
  created_at: string
}

export interface TicketTransition {
  id: string
  from_status: string
  to_status: string
  actor_id: string
  comment_id: string | null
  created_at: string
}

export interface TicketAssignment {
  id: string
  assigner_id: string
  assignee_id: string
  resulting_status: string
  context: {
    assignee_skills: string[]
    assignee_open_tickets: number
    ticket_priority: string
    ticket_severity: string
  }
  created_at: string
}

export interface TicketRequester {
  id: string
  name: string
  is_encargado: boolean
}

export interface RelatedFromItem {
  id: string
  ticket_number: string
  title: string
  record_type: 'Ticket' | 'Tarea'
}

/** Skill requerido para resolver el ticket, opcional (spec 011). */
export interface TicketSkillRef {
  id: string
  code: string
  label: string
}

export interface TicketDetail extends TicketListItem {
  description: string
  tool_id: string | null
  process_id: string | null
  estimated_resolution_minutes: number | null
  resolution_type_id: string | null
  related_ticket_id: string | null
  /** Registros (Ticket o Tarea) que referencian a este como "Registro relacionado" (Fase 3). */
  related_from: RelatedFromItem[]
  created_by: string
  /** Usuario/cliente solicitante asignado manualmente (Fase 2.2) — `null` si no hay o si `requester`
   * se resuelve automáticamente del creador (Fase 2.1). Editable solo cuando no es `null` o
   * cuando `requester` no viene de un creador con rol Usuario/cliente (ver TicketDetailPage). */
  client_contact_id: string | null
  requester: TicketRequester | null
  resolved_at: string | null
  resolution_accepted_at: string | null
  closed_at: string | null
  locked_fields: string[]
  close_eligible: boolean
  valid_actions: string[]
  comments: TicketComment[]
  transitions: TicketTransition[]
  assignments: TicketAssignment[]
  /** Subtareas (Nivel 5) de esta Tarea — vacío para Ticket y para Subtarea (spec 009). */
  subtasks: TicketListItem[]
  /** Skills requeridas para resolverlo, opcional y editable en cualquier estado (spec 011). */
  skills: TicketSkillRef[]
  /** Estado de SLA (Fase 4, spec 014) — 'sin_sla' para Tareas/Subtareas (FR-012). */
  sla: TicketSlaState
}

export interface TicketFormData {
  title: string
  description: string
  /** Requeridos para todos los roles salvo Usuario/cliente (alta simplificada, Fase 2.1 US3):
   * el backend los completa automáticamente (incident/medium/s3 + cliente del Usuario/cliente). */
  ticket_type?: TicketType
  priority?: Priority
  severity?: Severity
  client_id?: string
  project_id?: string | null
  /** Usuario/cliente solicitante (Fase 2.2) — opcional, debe pertenecer al `client_id` elegido. */
  client_contact_id?: string | null
  tool_id?: string | null
  process_id?: string | null
  record_type_id?: string | null
  escalation_level?: EscalationLevel
  related_ticket_id?: string | null
  /** Lista real (spec 009) — solo tiene efecto en una Tarea. */
  list_id?: string | null
  /** Marca el registro como Subtarea de la Tarea indicada (spec 009, Nivel 5). */
  parent_task_id?: string | null
  /** Encargado de la Tarea/Subtarea (spec 009) — opcional, default: el propio creador. */
  assignee_id?: string | null
  /** Skills requeridas al crear (spec 011) — se aplican tras el `create()` vía el endpoint
   * dedicado `PATCH /api/tickets/{id}/skills` (research.md Decisión 2); no viaja en el POST. */
  skill_ids?: string[]
}

export interface TicketFilters {
  page?: number
  page_size?: number
  search?: string
  client_id?: string
  project_id?: string
  status?: TicketStatus[]
  priority?: Priority
  severity?: Severity
  ticket_type?: TicketType
  assignee_id?: string
  escalation_level?: EscalationLevel
  sort?: string
  /** Fase 4, spec 014 — indicadores agregados de SLA (Historia 3). */
  sla_status?: TicketSlaState['status']
  sla_expiring_within_hours?: number
}

export interface PanelRow {
  resource: ResourceRef
  counts: Partial<Record<TicketStatus, number>>
  total: number
}

export interface PanelData {
  matrix: PanelRow[]
  unassigned_new: Array<{
    id: string
    ticket_number: string
    title: string
    priority: Priority
    severity: Severity
    client: EntityRef | null
    created_at: string
  }>
  statuses: TicketStatus[]
  status_labels: Record<string, string>
}

File: backend/domain/entities/ticket.py
from dataclasses import dataclass, field
from datetime import datetime
from typing import Optional
import uuid

# ── Catálogos fijos (CHECKs en DB) ─────────────────────────────────────

STATUSES = (
    "nuevo", "pre_analisis", "contacto", "en_analisis", "en_ejecucion",
    "en_pruebas", "pendiente_usuario", "resuelto", "cerrado", "cancelado",
)
FINAL_STATUSES = ("cerrado", "cancelado")
TICKET_TYPES = ("incident", "evolutive", "preventive")
PRIORITIES = ("critical", "high", "medium", "low")
SEVERITIES = ("s1", "s2", "s3", "s4")
ESCALATION_LEVELS = ("n1", "n2", "n3", "n4")

STATUS_LABELS = {
    "nuevo": "Nuevo", "pre_analisis": "Pre-Análisis", "contacto": "Contacto",
    "en_analisis": "En Análisis", "en_ejecucion": "En Ejecución",
    "en_pruebas": "En Pruebas", "pendiente_usuario": "Pendiente de Usuario",
    "resuelto": "Resuelto", "cerrado": "Cerrado", "cancelado": "Cancelado",
}

# Bloqueo de campos por estado (FR-010) — única fuente de verdad.
# Campos listados = NO editables via PATCH mientras el ticket esté en ese estado.
_BASE_LOCKED = {"status", "ticket_number", "record_type_id", "created_by"}
FIELD_LOCKS: dict[str, set[str]] = {
    "nuevo": set(),
    "pre_analisis": set(),
    "contacto": {"estimated_resolution_minutes", "severity", "priority"},
    "en_analisis": set(),  # se desbloquean tiempo estimado, severidad y prioridad
    "en_ejecucion": {"estimated_resolution_minutes"},
    "en_pruebas": {"estimated_resolution_minutes"},
    "pendiente_usuario": {"estimated_resolution_minutes", "severity", "priority"},
    "resuelto": {"estimated_resolution_minutes", "severity", "priority"},
    "cerrado": {"title", "description", "ticket_type", "priority", "severity",
                "escalation_level", "estimated_resolution_minutes"},
    "cancelado": {"title", "description", "ticket_type", "priority", "severity",
                  "escalation_level", "estimated_resolution_minutes"},
}


def locked_fields_for(status: str) -> list[str]:
    return sorted(_BASE_LOCKED | FIELD_LOCKS.get(status, set()))


def format_ticket_number(n: int) -> str:
    return f"TK-{n:06d}"


@dataclass
class Ticket:
    id: uuid.UUID
    ticket_number: int
    title: str
    description: str
    ticket_type: str
    priority: str
    severity: str
    client_id: uuid.UUID
    created_by: uuid.UUID
    record_type_id: Optional[uuid.UUID] = None
    status: str = "nuevo"
    escalation_level: str = "n2"
    project_id: Optional[uuid.UUID] = None
    tool_id: Optional[uuid.UUID] = None
    process_id: Optional[uuid.UUID] = None
    assignee_id: Optional[uuid.UUID] = None
    estimated_resolution_minutes: Optional[int] = None
    resolution_type_id: Optional[uuid.UUID] = None
    related_ticket_id: Optional[uuid.UUID] = None
    resolved_at: Optional[datetime] = None
    resolution_accepted_at: Optional[datetime] = None
    closed_at: Optional[datetime] = None
    created_at: datetime = field(default_factory=datetime.utcnow)
    updated_at: datetime = field(default_factory=datetime.utcnow)

    @property
    def number_display(self) -> str:
        return format_ticket_number(self.ticket_number)

    @property
    def is_final(self) -> bool:
        return self.status in FINAL_STATUSES

    def locked_fields(self) -> list[str]:
        return locked_fields_for(self.status)
